src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef CROSSLOADERALLOCATORHASH_INL
+#define CROSSLOADERALLOCATORHASH_INL
+#ifdef CROSSLOADERALLOCATORHASH_H
+#ifndef CROSSGEN_COMPILE
+
+#include "gcheaphashtable.inl"
+
+template <class TKey_, class TValue_>
+/*static*/ DWORD NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::ComputeUsedEntries(OBJECTREF &keyValueStore, DWORD *pEntriesInArrayTotal)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    DWORD entriesInArrayTotal = (((I1ARRAYREF)keyValueStore)->GetNumComponents() - sizeof(TKey))/sizeof(TValue);
+    DWORD usedEntries;
+    TValue* pStartOfValuesData = (TValue*)(((I1ARRAYREF)keyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
+
+    if (entriesInArrayTotal == 0)
+    {
+        usedEntries = 0;
+    }
+    else if ((entriesInArrayTotal >= 2) && (pStartOfValuesData[entriesInArrayTotal - 2] == (TValue)0))
+    {
+        usedEntries = (DWORD)pStartOfValuesData[entriesInArrayTotal - 1];
+    }
+    else if (pStartOfValuesData[entriesInArrayTotal - 1] == (TValue)0)
+    {
+        usedEntries = entriesInArrayTotal - 1;
+    }
+    else
+    {
+        usedEntries = entriesInArrayTotal;
+    }
+
+    *pEntriesInArrayTotal = entriesInArrayTotal;
+    return usedEntries;
+}
+
+#ifndef DACCESS_COMPILE
+template <class TKey_, class TValue_>
+/*static*/ void NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::SetUsedEntries(TValue* pStartOfValuesData, DWORD entriesInArrayTotal, DWORD usedEntries)
+{
+    if (usedEntries < entriesInArrayTotal)
+    {
+        if (usedEntries == (entriesInArrayTotal - 1))
+        {
+            pStartOfValuesData[entriesInArrayTotal - 1] = (TValue)0;
+        }
+        else
+        {
+            pStartOfValuesData[entriesInArrayTotal - 1] = (TValue)(usedEntries);
+            pStartOfValuesData[entriesInArrayTotal - 2] = (TValue)0;
+        }
+    }
+}
+
+template <class TKey_, class TValue_>
+/*static*/ void NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::AddToValuesInHeapMemory(OBJECTREF &keyValueStore, OBJECTREF &newKeyValueStore, const TKey& key, const TValue& value)
```

Comment:
We don't need the complication of two parameters. Returning a bool is just simpler. The reason it was there was to avoid any need to allocate a new GC_PROTECT region in case the target function needs to do a GC triggering operation while holding onto a temporary, but that isn't actually necessary at all. So, I've removed the extra parameter, and converted it to a pointer, which looks much better.